### PR TITLE
Search branch

### PR DESCRIPTION
**Pull request by ballparts**
Add search functionality for replays in menu.

### CODE REVIEW

File: src/js/modules/filter.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    // other quoted search terms
+    let quotedTerms = text.match(/".*?"/g) || [];
+    quotedTerms = quotedTerms.map(term => {
+      text = text.replace(term, '');
+      return term.replace(/"/g, '');
+    });
+
+    // non-quoted search terms preceded by hyphen
+    // should not identify terms containing hyphens in any but the first position
+    let negativeTerms = text.match(/^-\S*(?=\s)| -\S*(?=\s)|^-\S*$| -\S*$/g) || [];
```
Comment (chrahunt) [inline src/js/modules/filter.js]: At this point it's just space-separated search terms right? Can we do something like

```js
const tokens = text.match(/\S+/g) || [];
const negativeTerms = tokens
  .filter(s => s.startsWith('-'))
  .map(s => s.slice(1))
  .concat(quotedNegatives);
const terms = tokens
  .filter(s => !s.startsWith('-'))
  .concat(quotedTerms);
```

File: src/html/menus.html [hunk 2]
```diff
@@ -8,10 +8,28 @@ <h4 class="modal-title">TagPro Replays</h4>
...
-          <a class="pull-right" id="tpr-settings" data-toggle="modal" data-target="#settingsContainer">
-            <span class="glyphicon glyphicon-cog"></span>
-          </a>
-          <h3>Available Replays</h3>
+          <div class="container-fluid">
+            <div class="row">
+              <div id="tpr-title-div" class="col-sm-10">
+                <h3>Available Replays</h3>
+              </div>
+              <div id="tpr-search-div" class="col-sm-11" style="display: none;">
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/html/menus.html]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/5f1bd18abb9a40e378f0b93f96c8b68e0d17156a)

File: src/html/menus.html [hunk 3]
```diff
@@ -8,10 +8,28 @@ <h4 class="modal-title">TagPro Replays</h4>
...
+                <input id="searchInput" class="form-control" type="search" placeholder="Search">
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/html/menus.html]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/a465297c01cec2eddcdb90fd41832acca3b1b0a1)

File: src/html/menus.html [hunk 4]
```diff
@@ -39,6 +57,9 @@ <h3>Available Replays</h3>
           <div id="noReplays" class="text-center">
             <p>You don't have any replays saved. Go record some!</p>
           </div>
+          <div id="noSearchResults" class="text-center">
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/html/menus.html]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/06b093ff59279f9c94d68c564954de4d13bf83e7)

File: src/js/background.js [hunk 5]
```diff
@@ -944,8 +945,12 @@ chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
     return true;
 
   } else if (method == 'replay.list') {
+    let query = message.query || "";
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/background.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/869b4cf8437285574479317f9afda6a03c80001d)

File: src/js/background.js [hunk 6]
```diff
@@ -944,8 +945,12 @@ chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
...
     logger.info('Received request for the replay list.');
-    get_all_replays_info().then((info) => {
+    get_all_replays_info()
+    .then((info) => {
+      return new Filter(info, query).filter()
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/background.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/8064badc61fc505ad7bdf41b67d33aafde4e75ef)

File: src/js/content.js [hunk 7]
```diff
@@ -616,6 +619,22 @@ function initMenu() {
     })
   });
 
+  // Listen for search sumissions
+  let search = new Search();
+  search.on('submit', () => {
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/content.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/c9b6d7f91ca596a6eb1eaf8ffe7be2e5aa6f7604)

File: src/js/content.js [hunk 8]
```diff
@@ -616,6 +619,22 @@ function initMenu() {
...
+    logger.info('Submitting search query.');
+    replay_table.collection.query = search.query;
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/content.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/4f23ce94dfa238bbf4ecd3188eaea69205a7dce0)

I had trouble deciding whether two methods were necessary (`filter_replays` and `reset`) or if `filter_replays` could do both with an extra parameter. I decided to write two to keep the functions of each obvious in their names and to allow for separate behavior in the future.

File: src/js/modules/filter.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
+const parser = require('search-query-parser');
+
+class Filter {
+  /**
+  * metadata from get_all_replays_info()
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/f284633ca63aaccebd8acc9c59dfd81e5a18a835)

File: src/js/modules/filter.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  * query is literal search query string
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/02b4b0fe6953461e8f69ceeeb922c230684c18be)

File: src/js/modules/filter.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    this.query = query;
+    this.keywords = ['map', 'player', 'name'];
+  }
+
+  filter() {
+    return new Promise((resolve, reject) => {
+
+      try {
+        this.queryObject = parser.parse(this.query, {keywords: this.keywords});
+        if(this.queryObject === "") resolve(this.metadata);
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/b62ec0de8a19b8ba324cdb407894a95b289f23ac)

[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/232a69b1ac341dad5c88d96a40917e1ccf5b6a64)

File: src/js/modules/filter.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+      
+        if(typeof(this.queryObject) === "string") this.queryObject = {text: this.queryObject};
+
+
+        let results = this.metadata.filter(this._filterReplay, this);
+        if(!results) reject();
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/00b20979ee1f66bd974eb74a3630d846d2186cd9)

File: src/js/modules/filter.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        resolve(results);
+      } catch(err) {
+        reject(err);
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/307bff94811c6e42187fd75a94f47d14a12edf83)

File: src/js/modules/filter.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    replayMetadata.text = `${replayMetadata.map} ${replayMetadata.name} ${replayMetadata.player}`; // makes searching all text simpler
+    return replayMetadata;
+  }
+
+  /**
+  * function to be used in filter method
+  */
+  _filterReplay(replayMetadata) {
+    let flatMetadata = this._flatten(replayMetadata);
+    let matched = true;
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/5df3b482753ecf8bc146f8bc472514c839ba74dc)

File: src/js/modules/filter.js [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    
+    // handle excludes
+    if(this.queryObject.exclude) {
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/cffb87d65316b75470013ef467c6af9175cc730d)

File: src/js/modules/filter.js [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+  * function to be used in filter method
+  */
+  _filterReplay(replayMetadata) {
+    let flatMetadata = this._flatten(replayMetadata);
+    let matched = true;
+    
+    // handle excludes
+    if(this.queryObject.exclude) {  
+      let excluded = Object.keys(this.queryObject.exclude).some(item => {
+        if(typeof(this.queryObject.exclude[item]) === "string")
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/057f072792576e3615a6ff785a04d2884e82e809)

File: src/js/modules/filter.js [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    // handle free text - terms will be split by white space, unless within double quotes.
+    // this will handle exclusions in free text as well
+    // They are then treated as AND, so all must match
+    if(this.queryObject.text) {
+      let textTerms = this._formatText(this.queryObject.text);
+
+      let excludedText = textTerms.exclude.some(term => {
+        return flatMetadata.text.toLowerCase().includes(term.toLowerCase());
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/717b1dd98c848838eaca5b752426ca64813729ef)

File: src/js/modules/filter.js (hunk 1 above)
Comment (ballparts) [reply src/js/modules/filter.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/bbaa64f25ace127a87edde7666332126a1a2156e)

File: src/js/modules/replay-collection.js [hunk 18]
```diff
@@ -14,6 +14,7 @@ class ReplayCollection extends EventEmitter {
     super();
     this._collection = [];
     this._collection_by_id = new Map();
+    this.query = "";
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/replay-collection.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/0f52b06890192a5833bca3d370be2a76ed973346)

File: src/js/modules/search.js [hunk 19]
```diff
@@ -0,0 +1,90 @@
+const $ = require('jquery');
+const EventEmitter = require('events');
+
+require('../../../node_modules/bootstrap/js/tooltip.js');
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/search.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/349ada3e4d989e7c5c05bafbac548adb94d45b97)

File: src/js/modules/search.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+require('../../../node_modules/bootstrap/js/popover.js');
+
+var logger = require('util/logger')('search');
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/search.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/90df82fe2799c1535dbae98cfbea9141e72aa329)

File: src/js/modules/search.js [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+ * @event Search#reset
+ */
+class Search extends EventEmitter {
+  /**
+   * Listen on the provided element for search query.
+   */
+  constructor() {
+    super();
+    this.input = $('#searchInput');
+    this.icon = $('#tpr-search');
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/search.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/d339c248724739b44785a1377c57c65a2aa3e917)

I think this is what you meant. If you meant to also change their IDs in the HTML from `div` to `container`, let me know.

File: src/js/modules/search.js [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    this.resetButton = $('#tpr-search-reset');
+    this.inputDiv = $('#tpr-search-div');
+    this.titleDiv = $('#tpr-title-div');
+    this.iconDiv = $('#tpr-search-icon-div');
+    this.info = $('#tpr-search-info');
+
+    const inputHTML = '<span>Use this search to filter your available replays.' +
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/js/modules/search.js]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/a5daa01a20a2a5495410b3f02683faa48c8c9392)

File: src/scss/menu.scss [hunk 23]
```diff
@@ -157,3 +165,54 @@
...
+    left:  15px;
+    padding: 7px;
+}
+
+.glyphicon.right-addon {
+    position: absolute;
+    right: 15px;
+    padding: 7px;
+}
+
```
[the remark this answers is not in the document]
Comment (ballparts) [reply src/scss/menu.scss]: Addressed [here](https://github.com/chrahunt/TagProReplays/pull/220/commits/4d68fb3804f7d7efb5ce686e1ef0362c9fb94df5)

I checked all the files in this PR, I believe.

**Pull request merged**